// dominant-color-converter/lib/converter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Vibrant = require("node-vibrant");
var validation_1 = require("./validation");
var Converter = /** @class */ (function () {
    function Converter(customColorList) {
        this.customColorList = customColorList;
        if (customColorList) {
            if (!validation_1.validateCustomColors(customColorList)) {
                throw new Error('CustomColor should be "#ffffff" form.');
            }
            this.customColorList = customColorList;
        }
    }
    Converter.prototype._getOriginalPalette = function (url) {
        var _a, _b, _c, _d, _e, _f;
        return __awaiter(this, void 0, void 0, function () {
            var vibrant, palette;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        vibrant = Vibrant.from(url);
                        return [4 /*yield*/, vibrant.getPalette()];
                    case 1:
                        palette = _g.sent();
                        return [2 /*return*/, {
                                vibrant: (_a = palette.Vibrant) === null || _a === void 0 ? void 0 : _a.hex,
                                muted: (_b = palette.Muted) === null || _b === void 0 ? void 0 : _b.hex,
                                darkVibrant: (_c = palette.DarkVibrant) === null || _c === void 0 ? void 0 : _c.hex,
                                darkMuted: (_d = palette.DarkMuted) === null || _d === void 0 ? void 0 : _d.hex,
                                lightVibrant: (_e = palette.LightVibrant) === null || _e === void 0 ? void 0 : _e.hex,
                                lightMuted: (_f = palette.LightMuted) === null || _f === void 0 ? void 0 : _f.hex,
                            }];
                }
            });
        });
    };
    Converter.prototype._compareColors = function (originalColors) {
        originalColors = originalColors.slice(1);
        var numOriginalColor = parseInt(originalColors, 16);
        var numCustomColorList = this.customColorList.map(function (customColor) {
            return parseInt(customColor.slice(1), 16);
        });
        var differences = numCustomColorList.map(function (customColor) {
            return Math.abs(numOriginalColor - customColor);
        });
        var minDiff = Math.min.apply(Math, differences);
        return this.customColorList[differences.indexOf(minDiff)];
    };
    Converter.prototype.convert = function (url) {
        return __awaiter(this, void 0, void 0, function () {
            var originalPalette, originalColors, convertedColors;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._getOriginalPalette(url)];
                    case 1:
                        originalPalette = _a.sent();
                        if (!this.customColorList) {
                            return [2 /*return*/, originalPalette];
                        }
                        originalColors = Object.values(originalPalette);
                        convertedColors = originalColors.map(function (color) {
                            return _this._compareColors(color);
                        });
                        return [2 /*return*/, {
                                vibrant: convertedColors[0],
                                muted: convertedColors[1],
                                darkVibrant: convertedColors[2],
                                darkMuted: convertedColors[3],
                                lightVibrant: convertedColors[4],
                                lightMuted: convertedColors[5],
                            }];
                }
            });
        });
    };
    return Converter;
}());
exports.default = Converter;
